File: src/csharp/examples/cudnn/Mnist.cs
using System;
using Alea.CUDA;
using NUnit.Framework;

namespace Tutorial.Cs.examples.cudnn
{
    public class Test
    {
        //[CudnnMnistTest]
        [Test]
        public static void MnistTest()
        {
            var worker = Worker.Default;
            
            using (var network = new Network(worker))
            {
                var conv1 = Layer.Conv1(worker);
                var conv2 = Layer.Conv2(worker);
                var ip1 = Layer.Ip1(worker);
                var ip2 = Layer.Ip2(worker);

                Console.WriteLine("Classifying....");
                var i1 = network.ClassifyExample(Data.FirstImage, conv1, conv2, ip1, ip2);
                var i2 = network.ClassifyExample(Data.SecondImage, conv1, conv2, ip1, ip2);
                var i3 = network.ClassifyExample(Data.ThirdImage, conv1, conv2, ip1, ip2);

                Console.WriteLine("\n==========================================================\n");
                Console.WriteLine("Result of Classification: {0}, {1}, {2}", i1, i2, i3);

                if ((i1 != 1) || (i2 != 3) || (i3 != 5))
                    Console.WriteLine("Test Failed!!");
                else
                    Console.WriteLine("Test Passed!!");

                Console.WriteLine("\n==========================================================\n");
                Assert.AreEqual(i1, 1);
                Assert.AreEqual(i2, 3);
                Assert.AreEqual(i3, 5);
            }
        }
        //[/CudnnMnistTest]
    }
}

File: src/fsharp/examples/cudnn/Mnist.fs
(*** hide ***)
module Tutorial.Fs.examples.cudnn.Mnist
open System
open System.IO
open Alea.CUDA
open Alea.CUDA.CULib
open Alea.CUDA.CULib.CUBLASInterop
open Alea.CUDA.CULib.CUDNNInterop
open NUnit.Framework
open Layer
open Network

(*** define:CudnnMnistTest ***)
let [<Test>] test() =
    if Worker.Default.Device.Arch.Number < 300 then
        Assert.Inconclusive("running cudnn need at least cuda device of arch 3.0")
    else
        try
            let worker = Worker.Default
            use network = new Network(worker)
            let conv1, conv2 = Layer.conv1 worker, Layer.conv2 worker
            let ip1, ip2 = Layer.ip1 worker, Layer.ip2 worker
    
            printfn "Classifying...."
            let i1, i2, i3 =
                network.ClassifyExample FirstImage conv1 conv2 ip1 ip2,
                network.ClassifyExample SecondImage conv1 conv2 ip1 ip2,
                network.ClassifyExample ThirdImage conv1 conv2 ip1 ip2

            printfn "\n==========================================================\n"
            printfn "Result of Classification: %A, %A, %A" i1 i2 i3
            if i1 <> 1 || i2 <> 3 || i3 <> 5
            then printfn "Test Failed!!"
            else printfn "Test Passed!!"
            printfn "\n==========================================================\n"
    
            Assert.AreEqual(i1, 1)
            Assert.AreEqual(i2, 3)
            Assert.AreEqual(i3, 5)

        with :? System.DllNotFoundException -> Assert.Inconclusive("You need set the environment for cudnn native library to do this test.")